File: DotNet/VB.NET/VS2010/Threading/ProcessingThread.vb
'/***************************************************************
'* Copyright 2008-2016 - Accusoft Corporation, Tampa Florida.   *
'* This sample code is provided to Accusoft licensees "as is"   *
'* with no restrictions on use or modification. No warranty for *
'* use of this sample code is provided by Accusoft.             *
'****************************************************************/
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports Accusoft.ImagXpressSdk
Namespace Threading

    Public Class ProcessorThread
        Private myReferenceToMyLabel As System.Windows.Forms.Label
        Private myReferenceToMyViewer As Accusoft.ImagXpressSdk.ImageXView
        Private myImageFileName As System.String
        Private myResize As System.Int32
        Private myImage As Accusoft.ImagXpressSdk.ImageX
        Private myProcessor As Accusoft.ImagXpressSdk.Processor
        Private ImagXpress1 As Accusoft.ImagXpressSdk.ImagXpress


        Shared Sub AccusoftError(ByVal e As Accusoft.ImagXpressSdk.ImagXpressException)
            MessageBox.Show(Nothing, e.Message, "ImageXPress Error " + e.Number.ToString)
        End Sub

        Public Sub New(ByVal daImageFileName As System.String, ByVal daViewer As Accusoft.ImagXpressSdk.ImageXView, ByVal daLabel As System.Windows.Forms.Label, ByVal daSize As System.Int32, ByVal ImagXpress1 As ImagXpress)
            Me.myReferenceToMyViewer = daViewer
            Me.myReferenceToMyLabel = daLabel
            Me.myImageFileName = CType(daImageFileName.Clone, String)
            Me.myResize = daSize
            Me.ImagXpress1 = ImagXpress1
        End Sub

        Private Delegate Sub SafeContextImageChanger(ByVal newImage As Accusoft.ImagXpressSdk.ImageX)

        Private Delegate Sub SafeContextLabelChanger(ByVal newText As String)

        Private Delegate Sub SafeContextZoomChanger(ByVal newZoomFitType As Accusoft.ImagXpressSdk.ZoomToFitType)

        Private Sub RealZoomToFitChangerInTheSafeContext(ByVal newZoomFitType As Accusoft.ImagXpressSdk.ZoomToFitType)
            myReferenceToMyViewer.ZoomToFit(newZoomFitType)
        End Sub

        Private Sub RealImageChangerInTheSafeContext(ByVal newImage As Accusoft.ImagXpressSdk.ImageX)
            Dim oldImage As Accusoft.ImagXpressSdk.ImageX = myReferenceToMyViewer.Image
            myReferenceToMyViewer.Image = newImage
            If Not (oldImage Is Nothing) Then
                oldImage.Dispose()
                oldImage = Nothing
            End If
            GC.Collect()
            System.Threading.Thread.Sleep(25)
        End Sub

        Private Sub RealLabelChangerInTheSafeContext(ByVal newText As String)
            myReferenceToMyLabel.Text = newText
        End Sub

        Private Sub ChangeLabelText(ByVal newText As String)
            Dim theArgs As Object() = {newText}
            myReferenceToMyLabel.Invoke(New SafeContextLabelChanger(AddressOf RealLabelChangerInTheSafeContext), theArgs)
        End Sub

        Private Sub ChangeImage(ByVal newImage As Accusoft.ImagXpressSdk.ImageX)
            Dim theArgs As Object() = {newImage.Copy}
            myReferenceToMyViewer.Invoke(New SafeContextImageChanger(AddressOf RealImageChangerInTheSafeContext), theArgs)
        End Sub

        Private Sub ChangeZoomToFitType(ByVal newZoomFitType As Accusoft.ImagXpressSdk.ZoomToFitType)
            Dim theArgs As Object() = {newZoomFitType}
            myReferenceToMyViewer.Invoke(New SafeContextZoomChanger(AddressOf RealZoomToFitChangerInTheSafeContext), theArgs)
        End Sub

        Public Sub Process()
            Dim imageOpened As Boolean = False
            Dim TryCount As Integer = 0
            While (Not imageOpened) AndAlso (TryCount < 25)
                Try
                    myImage = Accusoft.ImagXpressSdk.ImageX.FromFile(ImagXpress1, myImageFileName)
                    imageOpened = True
                Catch ex As ImagXpressException
                    ChangeLabelText("Error opening image - " + ex.Number.ToString + ". Retrying...")
                End Try
                System.Math.Min(System.Threading.Interlocked.Increment(TryCount), TryCount - 1)
            End While
            If TryCount >= 25 Then
                Throw New Exception("Unable to open the image for processing.")
            End If
            System.Threading.Thread.Sleep(10)
            Try
                myProcessor = New Accusoft.ImagXpressSdk.Processor(ImagXpress1, myImage)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            Try
                ChangeImage(myImage)
                ChangeZoomToFitType(Accusoft.ImagXpressSdk.ZoomToFitType.FitBest)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            ChangeLabelText("Processing...")
            System.Threading.Thread.Sleep(10)
            ChangeLabelText("Resizing...")
            Try
                myProcessor.Resize(New System.Drawing.Size(myImage.ImageXData.Width * (myResize), myImage.ImageXData.Height * myResize), Accusoft.ImagXpressSdk.ResizeType.Quality)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            Try
                ChangeImage(myImage)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            ChangeLabelText("Blurring...")
            Try
                myProcessor.Blur()
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            Try
                ChangeImage(myImage)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            ChangeLabelText("Negating...")
            Try
                myProcessor.Negate()
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            Try
                ChangeImage(myImage)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            ChangeLabelText("Zooming...")
            Try
                ChangeZoomToFitType(Accusoft.ImagXpressSdk.ZoomToFitType.FitWidth)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            Try
                ChangeImage(myImage)
                ChangeZoomToFitType(Accusoft.ImagXpressSdk.ZoomToFitType.FitBest)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            ChangeLabelText("(Un)Negating...")
            Try
                myProcessor.Negate()
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            Try
                ChangeImage(myImage)
            Catch ex As ImagXpressException
                AccusoftError(ex)
            End Try
            System.Threading.Thread.Sleep(10)
            ChangeLabelText("Processed.")
            System.Threading.Thread.Sleep(10)
            myProcessor.Dispose()
            myProcessor = Nothing
            myImage.Dispose()
            myImage = Nothing
            GC.Collect()
            System.Threading.Thread.Sleep(100)
        End Sub
    End Class
End Namespace

File: DotNet/C#/VS2012/Threading/ProcessingThread.cs
/***************************************************************
* Copyright 2008-2016 - Accusoft Corporation, Tampa Florida.   *
* This sample code is provided to Accusoft licensees "as is"   *
* with no restrictions on use or modification. No warranty for *
* use of this sample code is provided by Accusoft.             *
****************************************************************/
using System;
using System.Windows.Forms;
using Accusoft.ImagXpressSdk;

namespace Threading
{
	/// <summary>
	/// Summary description for ProcessingThread
	/// </summary>
	public class ProcessorThread
	{
		//changing any of the ReferenceToMy stuff is not allowed. In order to properly manipulate
		private System.Windows.Forms.Label myReferenceToMyLabel;
		private Accusoft.ImagXpressSdk.ImageXView myReferenceToMyViewer; 
		//these, you need to use Invoke. The reason is that these objects are owned by a different thread

		private System.String myImageFileName;
		private System.Int32 myResize;
		private Accusoft.ImagXpressSdk.ImageX myImage;
		private Accusoft.ImagXpressSdk.Processor myProcessor;
		private ImagXpress daImagXpress;

		static void AccusoftError(Accusoft.ImagXpressSdk.ImagXpressException e) 
		{
			MessageBox.Show(null,e.Message,"ImageXPress Error " + e.Number.ToString());
		}

		public ProcessorThread (System.String daImageFileName,
			Accusoft.ImagXpressSdk.ImageXView daViewer,
			System.Windows.Forms.Label daLabel,
			System.Int32 daSize, ImagXpress daImagXpress)
		{			
			//set our references to these objects
			this.myReferenceToMyViewer = daViewer;
			this.myReferenceToMyLabel = daLabel;
			this.daImagXpress = daImagXpress;
			//since we do not need to pass back the filename, we can clone it to eliminate cross thread fears
			this.myImageFileName = (string)daImageFileName.Clone(); 
			
			//types such as System.Int32 are not real objects, so no cross thread fears.
			this.myResize = daSize;
		}

		private delegate void SafeContextImageChanger(Accusoft.ImagXpressSdk.ImageX newImage);

		private delegate void SafeContextLabelChanger(string newText);

		private delegate void SafeContextZoomChanger(Accusoft.ImagXpressSdk.ZoomToFitType newZoomFitType);

		private void RealZoomToFitChangerInTheSafeContext(Accusoft.ImagXpressSdk.ZoomToFitType newZoomFitType)
		{
			myReferenceToMyViewer.ZoomToFit(newZoomFitType);
		}

		private void RealImageChangerInTheSafeContext(Accusoft.ImagXpressSdk.ImageX newImage)
		{
			Accusoft.ImagXpressSdk.ImageX oldImage = myReferenceToMyViewer.Image;
	
			myReferenceToMyViewer.Image = newImage;

			if (oldImage != null)
			{
				oldImage.Dispose();
				oldImage = null;
			}
			GC.Collect();
			System.Threading.Thread.Sleep(25);
		}

		private void RealLabelChangerInTheSafeContext(string newText)
		{
			myReferenceToMyLabel.Text = newText;
		}

		private void ChangeLabelText(string newText)
		{
			object[] theArgs = {newText};
			myReferenceToMyLabel.Invoke(new SafeContextLabelChanger(this.RealLabelChangerInTheSafeContext),theArgs);
		}

		private void ChangeImage(Accusoft.ImagXpressSdk.ImageX newImage)
		{
			object[] theArgs = {newImage.Copy()};
			myReferenceToMyViewer.Invoke(new SafeContextImageChanger(this.RealImageChangerInTheSafeContext),theArgs);
		}

		private void ChangeZoomToFitType(Accusoft.ImagXpressSdk.ZoomToFitType newZoomFitType)
		{
			object[] theArgs = {newZoomFitType};
			myReferenceToMyViewer.Invoke(new SafeContextZoomChanger(this.RealZoomToFitChangerInTheSafeContext),theArgs);
		}

		public void Process()
		{	
			bool imageOpened = false;
			int TryCount = 0;
			while((!imageOpened) && (TryCount < 25))
			{
				try 
				{
					//Create the image object and open the image
					myImage = Accusoft.ImagXpressSdk.ImageX.FromFile(daImagXpress, myImageFileName);
					imageOpened = true;
				} 
				catch (ImagXpressException ex) 
				{
					ChangeLabelText("Error opening image - " + ex.Number.ToString() + ". Retrying...");
				}
				TryCount++;
			}

			if (TryCount >= 25)
			{
				throw new Exception("Unable to open the image for processing.");
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			try
			{
				//Create the processor object and tie it to the image we just opened
				myProcessor = new Accusoft.ImagXpressSdk.Processor(daImagXpress, myImage);
			} 
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			try 
			{
				//Set the viewer's image object equal to the one we are operating on...
				//we have to do this each time the image changes because we are making
				//copies of the image each time
				ChangeImage(myImage);

				//Set the viewing zoom type
				ChangeZoomToFitType(Accusoft.ImagXpressSdk.ZoomToFitType.FitBest);
			} 
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			//Change the viewer's label.
			ChangeLabelText("Processing...");

			//Rest
			System.Threading.Thread.Sleep(10);

			//Change the viewer's label.
			ChangeLabelText("Resizing...");

			try
			{
				//Resize the image
				myProcessor.Resize(new System.Drawing.Size(myImage.ImageXData.Width * (myResize), myImage.ImageXData.Height * myResize), Accusoft.ImagXpressSdk.ResizeType.Quality);
			}
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			try 
			{
				//Set the viewer's image object equal to the one we are operating on...
				//we have to do this each time the image changes because we are making
				//copies of the image each time
				ChangeImage(myImage);
			} 
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			//Change the viewer's label.
			ChangeLabelText("Blurring...");

			try
			{
				//Resize the image
				myProcessor.Blur();
			}
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			try 
			{
				//Set the viewer's image object equal to the one we are operating on...
				//we have to do this each time the image changes because we are making
				//copies of the image each time
				ChangeImage(myImage);
			} 
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			//Change the viewer's label.
			ChangeLabelText("Negating...");

			try
			{
				//Resize the image
				myProcessor.Negate();
			}
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			try 
			{
				//Set the viewer's image object equal to the one we are operating on...
				//we have to do this each time the image changes because we are making
				//copies of the image each time
				ChangeImage(myImage);
			} 
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			//Change the viewer's label.
			ChangeLabelText("Zooming...");

			try 
			{
				//Set the viewing zoom type
				ChangeZoomToFitType(Accusoft.ImagXpressSdk.ZoomToFitType.FitWidth);
			}
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			try 
			{
				//Set the viewer's image object equal to the one we are operating on...
				//we have to do this each time the image changes because we are making
				//copies of the image each time
				ChangeImage(myImage);

				//Set the viewing zoom type
				ChangeZoomToFitType(Accusoft.ImagXpressSdk.ZoomToFitType.FitBest);
			} 
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			//Change the viewer's label.
			ChangeLabelText("(Un)Negating...");

			try
			{
				//Resize the image
				myProcessor.Negate();
			}
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			try 
			{
				//Set the viewer's image object equal to the one we are operating on...
				//we have to do this each time the image changes because we are making
				//copies of the image each time
				ChangeImage(myImage);
			} 
			catch (ImagXpressException ex) 
			{
				AccusoftError(ex);
			}

			//Rest
			System.Threading.Thread.Sleep(10);

			//Change the viewer's label.
			ChangeLabelText("Processed.");
          
			//Rest
			System.Threading.Thread.Sleep(10);

			//kill the processor
			myProcessor.Dispose();
			myProcessor = null;

			//kill the local copy of the image
			myImage.Dispose();
			myImage = null;

			GC.Collect();
			System.Threading.Thread.Sleep(100);
		}
	}
}
